src/report_reader.cpp

```diff
@@ -374,43 +374,34 @@ DataFrame<T> ReportReader<T>::Population::get(const nonstd::optional<Selection>&
     }
 
     // Fill .data member
-
-    auto n_time_entries = index_stop - index_start + 1;
-    auto n_ids = data_frame.ids.size();
+    size_t n_time_entries = ((index_stop - index_start) / stride) + 1;
+    size_t n_ids = data_frame.ids.size();
     data_frame.data.resize(n_time_entries * n_ids);
 
-    // FIXME: It will be good to do it for ranges but if node_ids are not sorted it is not easy
-    // TODO: specialized this function for sorted node_ids?
-    int ids_index = 0;
-    for (const auto& node_id : node_ids) {
-        const auto it = std::find_if(
-            nodes_pointers_.begin(),
-            nodes_pointers_.end(),
-            [&node_id](const std::pair<NodeID, std::pair<uint64_t, uint64_t>>& node_pointer) {
-                return node_pointer.first == node_id;
-            });
-        if (it == nodes_pointers_.end()) {
-            continue;
-        }
-
-        // elems are by timestamp and by Nodes_id
-        std::vector<std::vector<float>> data;
-        pop_group_.getDataSet("data")
-            .select({index_start, it->second.first},
-                    {index_stop - index_start + 1, it->second.second - it->second.first})
-            .read(data);
-
-        int timer_index = 0;
-
-        for (const std::vector<float>& datum : data) {
-            std::copy(datum.data(),
-                      datum.data() + datum.size(),
-                      &data_frame.data[timer_index * n_ids + ids_index]);
-            ++timer_index;
+    std::vector<float> buffer(max - min);
+    auto dataset = pop_group_.getDataSet("data");
+    for (size_t timer_index = index_start; timer_index <= index_stop; timer_index += stride) {
+        dataset.select({timer_index, min}, {1, max - min}).read(buffer.data());
```

Comment:
Nice solution. This should work well for contiguous data, or horizontal chunks, while I don't expect good performance with vertical chunks. If this is the case (it could be tested), maybe it's possible to add a comment here or in the documentation?

Reply:
The solution is meant for the current column-based GID access, and it is supposed to avoid any changes in the HDF5 files (i.e., no chunking or any other external trick). The reader is meant to be as generic as possible for any kind of application, including for visualization purposes where vertical chunking would affect their performance.

We will discuss these matters in detail during our presentation, but thanks for the insight!

Reply:
> This should work well for contiguous data, or horizontal chunks, while I don't expect good performance with vertical chunks.

@GianlucaFicarelli : In [this plot](https://bbpteam.epfl.ch/project/issues/browse/BLPY-217?focusedCommentId=127949&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-127949), selecting single gid with all timeframes is vertical selection and performance is acceptable. Does that answer your question?

Obviously there are other improvements we would like to do but but those will come in separate PR.

Reply:
@pramodk I was meaning something different, I'll try to explain better: the same data can be stored in a h5 file contiguously, or in chunks of different shape and size.
1. If data is contiguous (and not transposed), then it's stored in C order (row-major) as (t0,gid0), (t0,gid1), (t0,gid2)..., and performance to read by rows with the solution in this PR is good.
2. If instead data is stored in vertical chunks (for example 1 gid and all its time frames in one chunk), then chunk 0 contains (t0,gid0), (t1,gid0), (t2,gid0)... In this case the solution of this PR has not good performance (I did some tests to confirm this).
3. Finally, with horizontal chunks (for example 1 chunk contains 1 time frame, and all the gids), the performance with the solution in this PR looks a bit worse than the contiguous case (i.e. 300 seconds vs 80 for 10K samples in my tests).
Anyway I think that the solution of this PR is good because it does not require any change to the contiguous h5 format that we are using.
I just think that it can be mentioned in the documentation that h5 chunking can affect performances.

Reply:
data are normalized: https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#frame-oriented-node-element-recordings

So they can be chunked, but not vertically oriented. This library only support this.

We have to find why this is worse if it is chunked.

Reply:
Ok, I now understand your point. 

> If instead data is stored in vertical chunks (for example 1 gid and all its time frames in one chunk), then chunk 0 contains (t0,gid0), (t1,gid0), (t2,gid0)... In this case the solution of this PR has not good performance (I did some tests to confirm this).

If the writer parts will be changed in the future, for sure we will need to reconsider optimisations. For example, if we chunk by column, the data for all time steps for specific gid will be read in one go and there won't be need to do filtering or finding min/max.

Reply:
#105 has been open about lower performances with chunked files.

Reply:
We added a comment in the source code, just to ensure future changes consider your suggestion, @GianlucaFicarelli . Thanks for the input!